drivers/metal/rendering_shader_container_metal.mm

```diff
@@ -0,0 +1,706 @@
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "rendering_shader_container_metal.h"
+#include "servers/rendering/rendering_device.h"
+
+#import "core/io/marshalls.h"
+
+#import <Metal/Metal.h>
+#import <spirv.hpp>
+#import <spirv_msl.hpp>
+#import <spirv_parser.hpp>
+
+Mutex MetalDeviceProfile::profiles_lock;
+HashMap<uint32_t, MetalDeviceProfile> MetalDeviceProfile::profiles;
+
+const MetalDeviceProfile *MetalDeviceProfile::get_profile(MetalDeviceProfile::Platform p_platform, MetalDeviceProfile::GPU p_gpu) {
+	DEV_ASSERT(p_platform == Platform::macOS || p_platform == Platform::iOS);
+
+	MutexLock lock(profiles_lock);
+
+	uint32_t key = (uint32_t)p_platform << 16 | (uint32_t)p_gpu;
+	if (MetalDeviceProfile *profile = profiles.getptr(key)) {
+		return profile;
+	}
+
+	MetalDeviceProfile res;
+	res.platform = p_platform;
+	res.gpu = p_gpu;
+	if (p_platform == Platform::macOS) {
+		res.features.mslVersionMajor = 3;
+		res.features.mslVersionMinor = 2;
+		res.features.argument_buffers_tier = ArgumentBuffersTier::Tier2;
+		res.features.simdPermute = true;
+	} else if (p_platform == Platform::iOS) {
+		switch (p_gpu) {
+			case GPU::Apple1:
+			case GPU::Apple2:
+			case GPU::Apple3:
+			case GPU::Apple4:
+			case GPU::Apple5: {
+				res.features.simdPermute = false;
+				res.features.argument_buffers_tier = ArgumentBuffersTier::Tier1;
+			} break;
+			case GPU::Apple6:
+			case GPU::Apple7:
+			case GPU::Apple8:
+			case GPU::Apple9: {
+				res.features.argument_buffers_tier = ArgumentBuffersTier::Tier2;
+				res.features.simdPermute = true;
+			} break;
+		}
+		res.features.mslVersionMajor = 3;
+		res.features.mslVersionMinor = 2;
+	}
+
+	return &profiles.insert(key, res)->value;
+}
+
+Error RenderingShaderContainerMetal::compile_metal_source(const char *p_source, const StageData &p_stage_data, Vector<uint8_t> r_binary_data) {
```

Comment:
```suggestion
Error RenderingShaderContainerMetal::compile_metal_source(const char *p_source, const StageData &p_stage_data, Vector<uint8_t> &r_binary_data) {
```

@DarioSamo `r_binary_data` should be a reference, I must have dropped it during some refactoring.